### PR TITLE
[#159519359] Remove conn retry on error 500 + Add a whitelist for errors to forward

### PR DESCRIPTION
**Pull request by LsKoder**
When PagoPaProxy provides a specific error 500, Backend uses a whitelist to check if the error should be forwarded to applicant

### CODE REVIEW

**Comment by digitalcitizenship** [comment]
## Affected stories

  * 🌟 [#159519359](https://www.pivotaltracker.com/story/show/159519359): Come APP, voglio che il backend ritorni lo stesso codice di stato HTTP ritornato da pagopa-proxy

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: src/services/pagoPAClientFactory.ts [hunk 1]
```diff
@@ -18,10 +19,15 @@ export default class PagoPAClientFactory
     codiceContestoPagamento: string,
     rptIdFromString: string
   ): ProxyPagoPA {
+    // Avoid to retry to send the request if proxy reply with a 500 HTTP code
+    const serviceClientOptions = {
+      noRetryPolicy: true
+    } as ServiceClientOptions;
```
Comment (cloudify) [inline src/services/pagoPAClientFactory.ts]: why the cast? try with the following instead:

```ts
const serviceClientOptions: ServiceClientOptions = {
      noRetryPolicy: true
    };
```

File: src/services/pagoPAClientFactory.ts [hunk 2]
```diff
@@ -18,10 +19,15 @@ export default class PagoPAClientFactory
     codiceContestoPagamento: string,
     rptIdFromString: string
   ): ProxyPagoPA {
+    // Avoid to retry to send the request if proxy reply with a 500 HTTP code
```
Comment (cloudify) [inline src/services/pagoPAClientFactory.ts]: I looks more like that with this option the client will NEVER retry a request, even for a timeout, is that the case?

File: src/services/pagoPAProxyService.ts [hunk 3]
```diff
@@ -105,3 +121,35 @@ export default class PagoPAProxyService {
...
+type ErrorBodyWithTitle = t.TypeOf<typeof ErrorBodyWithTitle>;
+
+/**
+ * Determine the error message to send to CDApp using a whitelist or a generic internal message
+ * Check Error Mapping documentation at:
+ * https://docs.google.com/document/d/1Qqe6mSfon-blHzc-ldeEHmzIkVaElKY5LtDnKiLbk80/edit#
+ */
+export function getErrorMessageForApp(
+  httpOperationResponse: HttpOperationResponse | undefined
+): string {
```
Comment (cloudify) [inline src/services/pagoPAProxyService.ts]:
```ts
ErrorBodyWithTitle.decode(httpOperationResponse)
  .map(_ => _.parsedBody.title)
  .filterOrElse(_ => pagoPaErrorWhiteList.indexOf(_) !== -1, internalErrorMessage)
```

Comment (cloudify) [reply src/services/pagoPAProxyService.ts]: https://github.com/gcanti/fp-ts/blob/master/src/Either.ts#L144

File: src/services/pagoPAClientFactory.ts (hunk 2 above)
Comment (LsKoder) [reply src/services/pagoPAClientFactory.ts]: As defined here:
https://github.com/Azure/ms-rest-js/blob/master/lib/serviceClient.ts
"if set to true, turn off the default retry policy."

I'm going to update the comment there.

File: src/services/pagoPAProxyService.ts (hunk 3 above)
Comment (LsKoder) [reply src/services/pagoPAProxyService.ts]: filterOrElse returns a Right<string> or a Left<ValidationError>

In this case I would like to return a string in both case (just select the string decoded or a default string).
To use filterOrElse I should pass a ValidationError as second parameter (internalErrorMessage is not allowed directly) and store the result to a const. Then, to extract again the value from this const.

Comment (cloudify) [reply src/services/pagoPAProxyService.ts]: then this:

```ts
ErrorBodyWithTitle.decode(httpOperationResponse)
  .map(_ => _.parsedBody.title)
  .filterOrElse(_ => pagoPaErrorWhiteList.indexOf(_) !== -1, internalErrorMessage)
  .getOrElse(internalErrorMessage)
```

Comment (LsKoder) [reply src/services/pagoPAProxyService.ts]: [ts] Argument of type '"Internal error."' is not assignable to parameter of type 'ValidationError[]'.

I can't pass internalErrorMessage as second parameter to filterOrElse. As I wrote in the last message, it requires ValidationError.

**Pull request merged**